### PR TITLE
[Doc] Add doc for Auto-Device Plugin 2021.4

### PR DESCRIPTION
**Pull request by coneypo**
### Details:
 - *Add AUTO.md for auto-device plugin*

### Tickets:
 - *57715*

### CODE REVIEW

File: docs/snippets/AUTO5.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include <ie_core.hpp>
+
+int main() {
+    std::string device_name = "MULTI:HDDL,GPU";
```
Comment (mashoujiang) [inline docs/snippets/AUTO5.cpp:4]: not MULTI..

File: docs/snippets/AUTO5.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    //! [part5]
+    InferenceEngine::Core ie; 
+    InferenceEngine::CNNNetwork cnnNetwork = ie.ReadNetwork("sample.xml");
+    // 'device_name' can be "AUTO:CPU,GPU" to configure the auto-device to use CPU and GPU
+    InferenceEngine::ExecutableNetwork exeNetwork = ie.LoadNetwork(cnnNetwork, device_name, full_config);
+    // new metric allows to query the optimization capabilities
+    std::vector<std::string> device_cap = exeNetwork.GetMetric(METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+    //! [part5]
+    return 0;
+}
```
Comment (mashoujiang) [inline docs/snippets/AUTO5.cpp:15]: this AUTO5.cpp is very similiar to AUTO3.cpp,  better to use another special usecase to distinguish them

**Comment by coneypo** [comment]
@myshevts  Hi, Maxim, please review this PR for auto-device plugin doc :)

**Comment by myshevts** [comment]
~ok for the initial ver (and to meet the code freeze)

**Review by myshevts — APPROVED** [review]
~ok for the

**Comment by myshevts** [comment]
@andrew-zaytsev / @ntyukaev pls review asap, let' try to meet the code-freeze

**Comment by coneypo** [comment]
Combined into one commit;

**Comment by mashoujiang** [comment]
depends on #6077 , could merge after #6077

**Comment by coneypo** [comment]
`InferenceEngine::ExecutableNetwork exec1 = ie.LoadNetwork(network);` in `/openvino/docs/snippets/AUTO0.cpp` depends on #6077, or CI will fail with `LoadNetwork(network)`

File: docs/snippets/AUTO0.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include <ie_core.hpp>
+
+int main() {
+//! [part0]
+    InferenceEngine::Core ie;
+    InferenceEngine::CNNNetwork network = ie.ReadNetwork("sample.xml");
+    //NEW IE-CENTRIC API, the "auto" plugin is (globally) pre-configured with the explicit option:
```
Comment (myshevts) [inline docs/snippets/AUTO0.cpp:7]: I see no global "pre-configuring" of the AUTO plugin in the example (presumably with ie.SetConfig(cong, "AUTO"))

Comment (myshevts) [reply docs/snippets/AUTO0.cpp:7]: this example seems to be also used below

File: docs/snippets/AUTO1.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include <ie_core.hpp>
+
+int main() {
+//! [part1]
+    InferenceEngine::Core ie;
+    InferenceEngine::CNNNetwork network = ie.ReadNetwork("sample.xml");
+    //NEW IE-CENTRIC API, the "AUTO" plugin is (globally) pre-configured with the explicit option:
```
Comment (myshevts) [inline docs/snippets/AUTO1.cpp:7]: I think I have added a change that explains that the 2 LoadNEtworks are equivalent and while first does rely on the pre-configuring the AUTO (see prev comment) the second does the conf for the specific ExecutableNetworkm via passing the device list to the LoadNetwork .

Notice that third option  is ie.LoadNetwork(network, "AUTO:CPU,GPU");

Comment (coneypo) [reply docs/snippets/AUTO1.cpp]: Got it, three options for ``ie.LoadNetwork`` there now;

**Review by myshevts — APPROVED** [review]

**Comment by myshevts** [comment]
pls also create the same PR, but to the release releases/2021/4 branch, asap, as the code freeze is coming.
@andrew-zaytsev @ntyukaev this is last call for the review

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+With the 2021.4 release, the Auto-Device internally recognizes and selects from CPU, 
+integrated GPU and discrete Intel GPUs (when available) depending on the device capabilities and the characteristic of CNN models, 
+for example, precisions. Then Auto-Device will assign inference requests to selected device.
+
+From the application point of view, this is just another device that handles all accelerators in full system. 
+
+With the 2021.4 release, the "setup" of Auto-device can be described in three major steps:
+* First is configuration of each device as usual (e.g. via conventional SetConfig method)
+* Second is loading of a network to the AUTO-Device plugin. This is the only change that you need in your application
+* Finally, just like with any other ExecutableNetwork (resulted from LoadNetwork) you just create as many requests as needed to saturate the devices.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:17]:
```suggestion
* Step 3: Just like with any other ExecutableNetwork (resulted from LoadNetwork), create as many requests as needed to saturate the devices. 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+integrated GPU and discrete Intel GPUs (when available) depending on the device capabilities and the characteristic of CNN models, 
+for example, precisions. Then Auto-Device will assign inference requests to selected device.
+
+From the application point of view, this is just another device that handles all accelerators in full system. 
+
+With the 2021.4 release, the "setup" of Auto-device can be described in three major steps:
+* First is configuration of each device as usual (e.g. via conventional SetConfig method)
+* Second is loading of a network to the AUTO-Device plugin. This is the only change that you need in your application
+* Finally, just like with any other ExecutableNetwork (resulted from LoadNetwork) you just create as many requests as needed to saturate the devices. 
+These steps are covered below in details
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:18]:
```suggestion
These steps are covered below in details.
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+
+## Defining and Configuring the Auto-Device plugin
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:21]:
```suggestion
## Defining and Configuring the Auto-Device Plugin
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+Following the OpenVINO notions of “devices”, the Auto-Device has a “AUTO” name. The only configuration option for AUTO-Device is a limited device list:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:22]:
```suggestion
Following the OpenVINO notions of “devices”, the Auto-device has “AUTO” name. The only configuration option for Auto-device is a limited device list:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+| Parameter name     | Parameter values      | Default            |             Description                                                      |
+| :---               | :---                  | :---               |:-----------------------------------------------------------------------------|
+| "AUTO_DEVICE_LIST" | comma-separated device names <span style="color:red">with no spaces</span>| N/A | Device candidate list to be selected    |
+
+You can use name of the configuration directly as a string, or use `IE::KEY_AUTO_DEVICE_LIST` from the `ie_plugin_config.hpp`,
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:28]:
```suggestion
You can use the configuration name directly as a string or use `IE::KEY_AUTO_DEVICE_LIST` from `ie_plugin_config.hpp`, 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+Following the OpenVINO notions of “devices”, the Auto-Device has a “AUTO” name. The only configuration option for AUTO-Device is a limited device list:
+
+| Parameter name     | Parameter values      | Default            |             Description                                                      |
+| :---               | :---                  | :---               |:-----------------------------------------------------------------------------|
+| "AUTO_DEVICE_LIST" | comma-separated device names <span style="color:red">with no spaces</span>| N/A | Device candidate list to be selected    |
+
+You can use name of the configuration directly as a string, or use `IE::KEY_AUTO_DEVICE_LIST` from the `ie_plugin_config.hpp`, 
+which defines the same string.
+
+Basically, there are two ways to use Auto-Device, directly indicate device by “AUTO” or empty string:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:31]:
```suggestion
There are two ways to use Auto-device:
1. Directly indicate device by “AUTO” or empty string:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+@snippet snippets/AUTO0.cpp part0
+
+Also, you can use Auto-Device configuration to limit the device candidate list to be selected:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:35]:
```suggestion
2. Use Auto-device configuration to limit the device candidates list to be selected:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+@snippet snippets/AUTO1.cpp part1
+
+Besides, Auto-Device supports query device optimization capabilities in metric;
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:39]:
```suggestion
Auto-device supports query device optimization capabilities in metric; 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+| Parameter name                 | Parameter values         |
+| :---                           | :---                     |
+| "OPTIMIZATION_CAPABILITIES"    | Auto-Device capabilities |
+
+## Enumerating Available Devices and Limit Auto Target Devices
+
+Inference Engine now features a dedicated API to enumerate devices and their capabilities. 
+See [Hello Query Device C++ Sample](../../../inference-engine/samples/hello_query_device/README.md).
+This is example output from the sample (truncated to the devices' names only):
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:49]:
```suggestion
This is the example output from the sample (truncated to the devices' names only):
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+```sh
+./hello_query_device
+Available devices: 
+    Device: CPU
+...
+    Device: GPU.0
+...
+    Device: GPU.1
+```
+Auto-device will select the most suitable device from available or limit devices to load,
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:60]:
```suggestion
Auto-device selects the most suitable device from available or limits devices to load, 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    Device: GPU.0
+...
+    Device: GPU.1
+```
+Auto-device will select the most suitable device from available or limit devices to load, 
+and simple programmatic way to load mode to device from available devices with the auto-device is as follows:
+
+@snippet snippets/AUTO2.cpp part2
+
+Another way to load mode to device from limit device choice with Auto-device::
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:65]:
```suggestion
Another way to load mode to device from limited choice of devices is with Auto-device::
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+@snippet snippets/AUTO3.cpp part3
+
+## Configuring the Individual Devices and Creating the Auto-Device On Top
+
+As discussed in the first section, you shall configure each individual device as usual and then just create the "AUTO" device on top:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:71]:
```suggestion
As described in the first section, configure each individual device as usual and then just create the "AUTO" device on top:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+@snippet snippets/AUTO4.cpp part4
+
+Alternatively, you can combine all the individual device settings into single config and load that, 
+allowing the Auto-device plugin to parse and apply that to the right devices. See code example here:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:76]:
```suggestion
allowing the Auto-device plugin to parse and apply it to the right devices. See the code example here:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+@snippet snippets/AUTO3.cpp part3
+
+## Configuring the Individual Devices and Creating the Auto-Device On Top
+
+As discussed in the first section, you shall configure each individual device as usual and then just create the "AUTO" device on top:
+
+@snippet snippets/AUTO4.cpp part4
+
+Alternatively, you can combine all the individual device settings into single config and load that,
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:75]:
```suggestion
Alternatively, you can combine all the individual device settings into a single config and load it, 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+@snippet snippets/AUTO4.cpp part4
+
+Alternatively, you can combine all the individual device settings into single config and load that, 
+allowing the Auto-device plugin to parse and apply that to the right devices. See code example here:
+
+@snippet snippets/AUTO5.cpp part5
+
+## Using the Auto-Device with OpenVINO Samples and Benchmark App
+
+Notice that every OpenVINO sample that supports "-d" (which stands for "device") command-line option transparently accepts the auto-device.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:82]:
```suggestion
Note that every OpenVINO sample that supports "-d" (which stands for "device") command-line option transparently accepts the auto-device. 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+Alternatively, you can combine all the individual device settings into single config and load that, 
+allowing the Auto-device plugin to parse and apply that to the right devices. See code example here:
+
+@snippet snippets/AUTO5.cpp part5
+
+## Using the Auto-Device with OpenVINO Samples and Benchmark App
+
+Notice that every OpenVINO sample that supports "-d" (which stands for "device") command-line option transparently accepts the auto-device. 
+The Benchmark Application is the best reference to the optimal usage of the auto-device.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:83]:
```suggestion
The Benchmark Application is the best example of the optimal usage of auto-device. 
```

Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:83]: Be consistent in capitalization of Auto-device. Use the same variant everywhere.

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+Alternatively, you can combine all the individual device settings into single config and load that, 
+allowing the Auto-device plugin to parse and apply that to the right devices. See code example here:
+
+@snippet snippets/AUTO5.cpp part5
+
+## Using the Auto-Device with OpenVINO Samples and Benchmark App
+
+Notice that every OpenVINO sample that supports "-d" (which stands for "device") command-line option transparently accepts the auto-device. 
+The Benchmark Application is the best reference to the optimal usage of the auto-device. 
+You don't need to setup number of requests, CPU threads as the application provides optimal out of the box performance.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:84]:
```suggestion
You do not need to set the number of requests and CPU threads, as the application provides optimal out-of-the-box performance. 
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+allowing the Auto-device plugin to parse and apply that to the right devices. See code example here:
+
+@snippet snippets/AUTO5.cpp part5
+
+## Using the Auto-Device with OpenVINO Samples and Benchmark App
+
+Notice that every OpenVINO sample that supports "-d" (which stands for "device") command-line option transparently accepts the auto-device. 
+The Benchmark Application is the best reference to the optimal usage of the auto-device. 
+You don't need to setup number of requests, CPU threads as the application provides optimal out of the box performance. 
+Below is example command-line to evaluate AUTO performance with that:
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:85]:
```suggestion
Below is the example command line to evaluate AUTO performance with that:
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+```sh
+./benchmark_app –d AUTO –m <model> -i <input> -niter 1000
+```
+You can also use the auto-device with limit device choice:
+
+```sh
+./benchmark_app –d AUTO:CPU,GPU –m <model> -i <input> -niter 1000
+```
+Note that default CPU stream is 1 if using “-d AUTO”.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:95]:
```suggestion
Note that the default CPU stream is 1 if using “-d AUTO”.
```

File: docs/IE_DG/supported_plugins/AUTO.md [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+Notice that you can use the FP16 IR to work with auto-device. Also notice that no demos are (yet) fully optimized for the auto-device, by means of selecting the most suitable device , using the GPU streams/throttling, and so on.
```
Comment (avladimi) [inline docs/IE_DG/supported_plugins/AUTO.md:97]:
```suggestion
Note that you can use the FP16 IR to work with auto-device. Also note that no demos are (yet) fully optimized for the auto-device, by means of selecting the most suitable device, using the GPU streams/throttling, and so on.
```

**Pull request merged**